test/hotspot/jtreg/compiler/loopopts/superword/TestMemorySegment.java

```diff
@@ -0,0 +1,860 @@
+//  * @bug 8310190
+//  * @summary Test vectorization of loops over MemorySegment
+//  * @library /test/lib /
+//  * @run driver compiler.loopopts.superword.TestMemorySegment Mixed
+//  */
+
+public class TestMemorySegment {
+    public static void main(String[] args) {
+        TestFramework framework = new TestFramework(TestMemorySegmentImpl.class);
+        framework.addFlags("-DmemorySegmentProviderNameForTestVM=" + args[0]);
+        framework.setDefaultWarmup(100);
+        framework.start();
+    }
+}
+
+class TestMemorySegmentImpl {
+    static final int BACKING_SIZE = 1024 * 8;
+    static final Random RANDOM = Utils.getRandomInstance();
+
+
+    interface TestFunction {
+        Object[] run();
+    }
+
+    interface MemorySegmentProvider {
+        MemorySegment newMemorySegment();
+    }
+
+    static MemorySegmentProvider provider;
+
+    static {
+        String providerName = System.getProperty("memorySegmentProviderNameForTestVM");
+        provider = switch (providerName) {
+            case "ByteArray"        -> ( () -> { return newMemorySegmentOfByteArray(); } );
+            case "CharArray"        -> ( () -> { return newMemorySegmentOfCharArray(); } );
+            case "ShortArray"       -> ( () -> { return newMemorySegmentOfShortArray(); } );
+            case "IntArray"         -> ( () -> { return newMemorySegmentOfIntArray(); } );
+            case "LongArray"        -> ( () -> { return newMemorySegmentOfLongArray(); } );
+            case "FloatArray"       -> ( () -> { return newMemorySegmentOfFloatArray(); } );
+            case "DoubleArray"      -> ( () -> { return newMemorySegmentOfDoubleArray(); } );
+            case "ByteBuffer"       -> ( () -> { return newMemorySegmentOfByteBuffer(); } );
+            case "ByteBufferDirect" -> ( () -> { return newMemorySegmentOfByteBufferDirect(); } );
+            case "Native"           -> ( () -> { return newMemorySegmentOfNative(); } );
+            case "MixedArray"       -> ( () -> { return newMemorySegmentOfMixedArray(); } );
+            case "MixedBuffer"      -> ( () -> { return newMemorySegmentOfMixedBuffer(); } );
+            case "Mixed"            -> ( () -> { return newMemorySegmentOfMixed(); } );
+            default -> throw new RuntimeException("Test argument not recognized: " + providerName);
+        };
+    }
+
+    // List of tests
+    Map<String,TestFunction> tests = new HashMap<String,TestFunction>();
+
+    // List of gold, the results from the first run before compilation
+    Map<String,Object[]> golds = new HashMap<String,Object[]>();
+
+    public TestMemorySegmentImpl () {
+        // Generate two MemorySegments as inputs
+        MemorySegment a = newMemorySegment();
+        MemorySegment b = newMemorySegment();
+        fillRandom(a);
+        fillRandom(b);
+
+        // Add all tests to list
+        tests.put("testMemorySegmentBadExitCheck", () -> {
+            return testMemorySegmentBadExitCheck(copy(a));
+        });
```

Comment:
Same as above, you can replace this with an expression lambda:
```suggestion
        tests.put("testIntLoop_longIndex_intInvar_sameAdr_byte", 
                  () -> testIntLoop_longIndex_intInvar_sameAdr_byte(copy(a), 0));
```
Same for others.

Reply:
Nice idea!